src/ansi-c/goto-conversion/goto_clean_expr.cpp

```diff
@@ -16,9 +16,172 @@ Author: Daniel Kroening, kroening@kroening.com
+  const code_expressiont &code,
+  const irep_idt &mode,
+  goto_convertt &converter)
+{
+  goto_programt where;
+  converter.goto_convert(code, where, mode);
+  where.compute_location_numbers();
+
+  natural_loops_mutablet natural_loops(where);
+  INVARIANT(
+    natural_loops.loop_map.size() == 0,
+    "quantifier must not contain side effects");
+
+  // `last` is the instruction corresponding to the last expression in the
+  // statement expression.
+  goto_programt::const_targett last;
+  for(goto_programt::const_targett it = where.instructions.begin();
+      it != where.instructions.end();
+      ++it)
+  {
+    // `last` is an other-instruction.
+    if(it->is_other())
+    {
+      last = it;
+    }
+  }
+  auto last_expr =
+    expr_try_dynamic_cast<code_expressiont>(last->get_other())->expression();
+
+  // `pathes` contains all the `targett` we are iterating over.
+  std::vector<goto_programt::const_targett> pathes;
+  pathes.push_back(where.instructions.begin());
+  std::vector<std::pair<exprt, replace_symbolt>> path_conditions_and_value_maps;
+  path_conditions_and_value_maps.push_back(
+    std::make_pair(true_exprt(), replace_symbolt()));
+
+  std::unordered_set<symbol_exprt, irep_hash> declared_symbols;
+  // All bounded variables are local.
+  declared_symbols.insert(qex.variables().begin(), qex.variables().end());
+
+  exprt res = true_exprt();
+
+  // Visit the quantifeir body along `pathes`.
+  while(!pathes.empty())
+  {
+    auto &current_it = pathes.back();
+    auto &path_condition = path_conditions_and_value_maps.back().first;
+    auto &value_map = path_conditions_and_value_maps.back().second;
+    INVARIANT(
+      current_it != where.instructions.end(),
+      "Quantifier body must have a unique end expression.");
+
+    // Add all local-declared symbols into `declared_symbols`.
+    if(current_it->is_decl())
+    {
+      declared_symbols.insert(current_it->decl_symbol());
+    }
+    // ASSIGN lhs := rhs
+    // Add the replace lhr <- value_map(rhs) to the current value_map.
+    if(current_it->is_assign())
+    {
+      // Check that if lhs is a declared symbol.
+      auto lhs = current_it->assign_lhs();
+      INVARIANT(
+        declared_symbols.count(find_base_symbol(lhs)),
+        "quantifier must not contain side effects");
+      exprt rhs = current_it->assign_rhs();
+      INVARIANT(
+        !has_subexpr(rhs, ID_function_call),
+        "quantifier must not contain side effects");
+      value_map.replace(rhs);
+      value_map.set(to_symbol_expr(current_it->assign_lhs()), rhs);
+    }
+    // GOTO label
+    // -----------
+    // Move the current targett to label.
+    // or
+    // IF cond GOTO label
+    // ----------
+    // Move the current targett to targett+1 with path condition
+    // path_condition && !cond;
+    // and add a new path starting from label with path condition
+    // path_condition && cond.
+    if(current_it->is_goto())
+    {
+      if(current_it->has_condition())
```

Comment:
I don't think this does what you want here: `has_condition()` will always be `true` when `is_goto()` is `true`. Perhaps you wanted `!current_it->condition().is_true()`?